### PR TITLE
Eliminate RepositoryPlugin.list_partitions (#6531)

### LINKED ISSUE #3579: Move concatenation of prefixes out of plugin
Currently plugin.list_bundles concatenates the common prefix and the partition prefix that is passed in.
This limits the utility of this method. The caller should combine the prefixes.

### LINKED ISSUE #6531: Eliminate `RepositoryPlugin.list_partitions`
https://github.com/DataBiosphere/azul/issues/6529#user-content-6531

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #6531

Note: this is to be promoted separately.

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] This PR is labeled `partial` <sub>or completely resolves all connected issues</sub>
- [x] This PR partially resolves each of the connected issues <sub>or does not have the `partial` label</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or is not chained to another PR</sub>
- [x] The blocking PR is labeled `base` <sub>or this PR is not chained to another PR</sub>
- [x] This PR is labeled `chained` <sub>or is not chained to another PR</sub>

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or the changes introduced by this PR will not require reindexing of any deployment</sub>
- [x] This PR is labeled `reindex:dev` <sub>or the changes introduced by it will not require reindexing of `dev`</sub>
- [x] This PR is labeled `reindex:anvildev` <sub>or the changes introduced by it will not require reindexing of `anvildev`</sub>
- [x] This PR is labeled `reindex:anvilprod` <sub>or the changes introduced by it will not require reindexing of `anvilprod`</sub>
- [x] This PR is labeled `reindex:prod` <sub>or the changes introduced by it will not require reindexing of `prod`</sub>
- [x] This PR is labeled `reindex:partial` and its description documents the specific reindexing procedure for `dev`, `anvildev`, `anvilprod` and `prod` <sub>or requires a full reindex or carries none of the labels `reindex:dev`, `reindex:anvildev`, `reindex:anvilprod` and `reindex:prod`</sub>
- [x] This PR and its connected issues are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (upgrading deployments)

- [x] Ran `make docker_images.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `docker_images.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the none of the stable branches (`anvilprod` and `prod`) have temporary hotfixes for any of the issues connected to this PR</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] PR is awaiting requested review from system administrator
- [x] PR is assigned to only the system administrator

### System administrator (after approval)

- [ ] ~~Actually approved the PR~~
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] A comment to this PR details the completed security design review
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved connected issues to *Approved* column
- [x] PR is assigned to only the operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex:…` labels and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>

### System administrator

- [x] Background migrations for `dev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for `anvildev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator

### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] **Performed equivalent reindex on `develop` to serve as a baseline** (`dev` Job [100952](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/100952), `anvildev` Job [48695](https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/48695))
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only included `p` if the PR is also labeled `partial`</sub>
- [x] Moved connected issues to *Merged lower* column in ZenHub
- [x] Moved blocked issues to *Triage* <sub>or no issues are blocked on the connected issues</sub>
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [ ] ~Build passes on GitLab `anvildev`~ https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/48994
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

### Operator (reindex)

- [x] Deindexed all unreferenced catalogs in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed all unreferenced catalogs in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Deindexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Indexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Indexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub> (Job [101121](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/101121))
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub> (Job [48855](https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/48855))
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>

### Operator

- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels to the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels, from the description of this PR to that of the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to no one

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: src/azul/plugins/metadata/hca/indexer/transform.py [hunk 1]
```diff
@@ -397,14 +392,14 @@ def __init__(self, id: str, title: str, category: SubmitterCategory):
         self.by_id[id] = self
 
     @classmethod
-    def for_id(cls, submitter_id: str) -> Optional['Submitter']:
+    def for_id(cls, submitter_id: str) -> 'Submitter | None':
```
Comment (dsotirho-ucsc) [inline src/azul/plugins/metadata/hca/indexer/transform.py]:
```suggestion
    def for_id(cls, submitter_id: str) -> Self | None:
```

File: src/azul/plugins/metadata/hca/indexer/transform.py [hunk 2]
```diff
@@ -397,14 +392,14 @@ def __init__(self, id: str, title: str, category: SubmitterCategory):
...
         try:
             return cls.by_id[submitter_id]
         except KeyError:
             return None
 
     @classmethod
-    def for_file(cls, file: api.File) -> Optional['Submitter']:
+    def for_file(cls, file: api.File) -> 'Submitter | None':
```
Comment (dsotirho-ucsc) [inline src/azul/plugins/metadata/hca/indexer/transform.py]:
```suggestion
    def for_file(cls, file: api.File) -> Self | None:
```

File: deployments/dev/environment.py [hunk 3]
```diff
@@ -64,121 +53,121 @@ def mkdict(previous_catalog: dict[str, str],
...
+    mksrc('bigquery', 'datarepo-dev-3d6d24ad', 'hca_dev_05657a599f9d4bb9b77b24be13aa5cea__20210827_20230215'),
+    mksrc('bigquery', 'datarepo-dev-b46086a9', 'hca_dev_51f02950ee254f4b8d0759aa99bb3498__20210827_20230215'),
+    mksrc('bigquery', 'datarepo-dev-27c50fbc', 'hca_dev_c5f4661568de4cf4bbc2a0ae10f08243__20210827_20230215'),
+    mksrc('bigquery', 'datarepo-dev-2a4ab485', 'hca_dev_d2111fac3fc44f429b6d32cd6a828267__20210830_20230215'),
+    mksrc('bigquery', 'datarepo-dev-1005632d', 'hca_dev_d3ac7c1b53024804b611dad9f89c049d__20210827_20230215'),
+    mksrc('bigquery', 'datarepo-dev-8709b362', 'hca_dev_df88f39f01a84b5b92f43177d6c0f242__20210827_20230215'),
+    mksrc('bigquery', 'datarepo-dev-e1712bfa', 'hca_dev_e526d91dcf3a44cb80c5fd7676b55a1d__20210902_20230215'),
 ]))
 
 lungmap_sources = mkdict({}, 2, mkdelta([
```
Comment (dsotirho-ucsc) [inline deployments/dev/environment.py:173]: The old `subgraphs` argument `1` needs to be removed from the `lungmap_sources` sources.

**Review by dsotirho-ucsc — CHANGES_REQUESTED** [review]

File: test/integration_test.py [hunk 4]
```diff
@@ -286,73 +287,17 @@ def managed_access_sources_by_catalog(self
...
-            # arbitrary.
-            upper = 64
-            lower = 1
-
-        self.assertLessEqual(num_bundles, upper, partition + ' is too large')
-        self.assertGreaterEqual(num_bundles, lower, partition + ' is too small')
-
-        return source, partition_prefix, fqids
-
-    def _list_bundles(self,
```
Comment (dsotirho-ucsc) [inline test/integration_test.py:349]: A comment in `_prepare_notifications()` that references `_list_bundles()` needs to be updated.

**Review by dsotirho-ucsc — CHANGES_REQUESTED** [review]

**Review by dsotirho-ucsc — DISMISSED** [review]
Approved.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Please separate the annotation commits, one for Optional, one for Set/AbstractSet, one for Type and so on as needed.

```diff
Index: src/azul/plugins/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/__init__.py b/src/azul/plugins/__init__.py
--- a/src/azul/plugins/__init__.py	(revision 10d286acf57927b4757171cd66b4055bd11743aa)
+++ b/src/azul/plugins/__init__.py	(date 1727114349843)
@@ -613,8 +613,9 @@
     @abstractmethod
     def _count_subgraphs(self, source: SOURCE_SPEC) -> int:
         """
-        Count the total number of subgraphs stored in the given source, ignoring
-        its prefix.
+        The total number of subgraphs in the given source, ignoring its prefix.
+
+        REVIEW: Zero seems to have special meaning, which should be documented.
         """
         raise NotImplementedError
 
@@ -634,9 +635,9 @@
             # We use the "other" heuristic during IT to avoid indexing an
             # excessive number of bundles
             if is_main and not is_it:
-                prefix = Prefix.for_main_source(count)
+                prefix = Prefix.for_main_deployment(count)
             else:
-                prefix = Prefix.for_other_source(count)
+                prefix = Prefix.for_lesser_deployment(count)
             source = attr.evolve(source, spec=attr.evolve(source.spec, prefix=prefix))
         return source
 
Index: src/azul/plugins/repository/canned/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/canned/__init__.py b/src/azul/plugins/repository/canned/__init__.py
--- a/src/azul/plugins/repository/canned/__init__.py	(revision 10d286acf57927b4757171cd66b4055bd11743aa)
+++ b/src/azul/plugins/repository/canned/__init__.py	(date 1727114148447)
@@ -164,6 +164,7 @@
             return factory.load_staging_area(path)
 
     def _count_subgraphs(self, source: SOURCE_SPEC) -> int:
+        # REVIEW: A comment should explain why 0 is a suitable value
         return 0
 
     def list_bundles(self,
Index: scripts/post_deploy_tdr.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/scripts/post_deploy_tdr.py b/scripts/post_deploy_tdr.py
--- a/scripts/post_deploy_tdr.py	(revision 10d286acf57927b4757171cd66b4055bd11743aa)
+++ b/scripts/post_deploy_tdr.py	(date 1726853400195)
@@ -58,8 +58,10 @@
         result = [
             catalog.name
             for catalog in config.catalogs.values()
-            if (config.is_tdr_enabled(catalog.name)
-                and catalog.name not in config.integration_test_catalogs)
+            if (
+                config.is_tdr_enabled(catalog.name)
+                and catalog.name not in config.integration_test_catalogs
+            )
         ]
         assert result, config.catalogs
         return result
Index: src/azul/plugins/repository/dss/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/dss/__init__.py b/src/azul/plugins/repository/dss/__init__.py
--- a/src/azul/plugins/repository/dss/__init__.py	(revision 10d286acf57927b4757171cd66b4055bd11743aa)
+++ b/src/azul/plugins/repository/dss/__init__.py	(date 1727114349847)
@@ -118,6 +118,7 @@
         return DSSSourceRef.id_from_spec(spec)
 
     def _count_subgraphs(self, source: SimpleSourceSpec) -> int:
+        # REVIEW: A comment should explain why 0 is a suitable value
         return 0
 
     def list_sources(self,
Index: src/azul/terra.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/terra.py b/src/azul/terra.py
--- a/src/azul/terra.py	(revision 10d286acf57927b4757171cd66b4055bd11743aa)
+++ b/src/azul/terra.py	(date 1726852810880)
@@ -139,7 +139,8 @@
         ...
         ValueError: 'eggs' is not a valid TDRSourceSpec.Domain
 
-        >>> # If any :'s are missing, the last part will be interpreted as the prefix
+        If any :'s are missing, the last part will be interpreted as the prefix
+
         >>> TDRSourceSpec.parse('tdr:bigquery:gcp:foo:bar')
         Traceback (most recent call last):
         ...
Index: src/azul/indexer/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/indexer/__init__.py b/src/azul/indexer/__init__.py
--- a/src/azul/indexer/__init__.py	(revision 10d286acf57927b4757171cd66b4055bd11743aa)
+++ b/src/azul/indexer/__init__.py	(date 1727130748825)
@@ -21,7 +21,8 @@
     TypedDict,
 )
 
-import attr
+# REVIEW: We shouldn't mix the old and the new API, converting a module to the
+#         new API should happen in a separate commit
 import attrs
 
 from azul import (
@@ -144,34 +145,36 @@
         return cls(common=entry, partition=partition)
 
     @classmethod
-    def for_main_source(cls, subgraphs: int) -> Self:
+    def for_main_deployment(cls, num_subgraphs: int) -> Self:
         """
         A prefix that is expected to rarely exceed 8192 subgraphs per partition
 
-        >>> str(Prefix.for_main_source(0))
+        >>> str(Prefix.for_main_deployment(0))
         Traceback (most recent call last):
         ...
         ValueError: math domain error
 
-        >>> str(Prefix.for_main_source(1))
+        >>> str(Prefix.for_main_deployment(1))
         '/0'
 
-        >>> n = 8192
         >>> cases = [-1, 0, 1, 2]
-        >>> [str(Prefix.for_main_source(n + i)) for i in cases]
+
+        >>> n = 8192
+        >>> [str(Prefix.for_main_deployment(n + i)) for i in cases]
         ['/0', '/0', '/1', '/1']
 
-        >>> # Sources with this many bundles are very rare, so we have a
-        >>> # generous margin of error surrounding this cutoff point
-        >>> n *= 16  # 131072
-        >>> [str(Prefix.for_main_source(n + i)) for i in cases]
+        Sources with this many bundles are very rare, so we have a generous
+        margin of error surrounding this cutoff point
+
+        >>> n = 8192 * 16
+        >>> [str(Prefix.for_main_deployment(n + i)) for i in cases]
         ['/1', '/1', '/2', '/2']
         """
-        partition = cls._prefix_length(subgraphs, 8192)
+        partition = cls._prefix_length(num_subgraphs, 8192)
         return cls(common='', partition=partition)
 
     @classmethod
-    def for_other_source(cls, subgraphs: int) -> Self:
+    def for_lesser_deployment(cls, num_subgraphs: int) -> Self:
         """
         A prefix that yields an average of approximately 24 subgraphs per
         source, using an experimentally derived heuristic formula designed to
@@ -179,26 +182,27 @@
         partition prefix length is always 1, even though some partitions may be
         empty, to provide test coverage for handling multiple partitions.
 
-        >>> str(Prefix.for_other_source(0))
+        >>> str(Prefix.for_lesser_deployment(0))
         Traceback (most recent call last):
         ...
         ValueError: math domain error
 
-        >>> str(Prefix.for_other_source(1))
+        >>> str(Prefix.for_lesser_deployment(1))
         '/1'
 
-        >>> n = 64
         >>> cases = [-1, 0, 1, 2]
-        >>> [str(Prefix.for_other_source(n + i)) for i in cases]
+
+        >>> n = 64
+        >>> [str(Prefix.for_lesser_deployment(n + i)) for i in cases]
         ['/1', '/1', '0/1', '1/1']
 
-        >>> n *= 16  # 1024
-        >>> [str(Prefix.for_other_source(n + i)) for i in cases]
+        >>> n = 64 * 16
+        >>> [str(Prefix.for_lesser_deployment(n + i)) for i in cases]
         ['e/1', 'f/1', '00/1', '10/1']
         """
-        digits = f'{subgraphs - 1:x}'[::-1]
-        length = cls._prefix_length(subgraphs, 64)
-        assert length < len(digits), subgraphs
+        digits = f'{num_subgraphs - 1:x}'[::-1]
+        length = cls._prefix_length(num_subgraphs, 64)
+        assert length < len(digits), num_subgraphs
         return cls(common=digits[:length], partition=1)
 
     @classmethod
@@ -269,6 +273,11 @@
     have simple unstructured names may want to use :class:`SimpleSourceSpec`.
     """
 
+    # PREVIEW: A comment should explain what the significance of None is and
+    #          how that differs from '', assuming we need the distinction. If we
+    #          don't, I would prefer to represent that case with '' instead of
+    #          None.
+    #
     prefix: Prefix | None
 
     @classmethod
@@ -300,7 +309,7 @@
         >>> p('foo:4/0').eq_ignoring_prefix(p('bar:4/0'))
         False
         """
-        return self == attr.evolve(other, prefix=self.prefix)
+        return self == attrs.evolve(other, prefix=self.prefix)
 
 
 @attrs.frozen(kw_only=True)
```

**Comment by nadove-ucsc** [comment]
> REVIEW: Zero seems to have special meaning, which should be documented.
> REVIEW: A comment should explain why 0 is a suitable value

I was using zero as a placeholder with the expectation that those overrides would never actually be called. It has no special meaning at the call site and would raise an exception if actually encountered there. I've replaced it with an actual implementation for the Canned plugin (untested as it's never called) and an assertion for the DSS plugin.

**Comment by nadove-ucsc** [comment]
> REVIEW: A comment should explain what the significance of None is and
              how that differs from '', assuming we need the distinction. If we
              don't, I would prefer to represent that case with '' instead of
              None.

This makes no sense to me. The `prefix` field can never be equal to `''` because the type annotation `Prefix | None` does not permit instances of `str`. Furthermore, [the empty string cannot be parsed as a Prefix](https://github.com/DataBiosphere/azul/blob/e8c2280039206d42dbd3623611f32c02fa16708b/src/azul/indexer/__init__.py#L126), and there is no instance of `class Prefix` whose string representation is `''`. The "distinction" is therefore that `None` is a valid possible value for this field and the empty string is not.

**Comment by nadove-ucsc** [comment]
> REVIEW: We shouldn't mix the old and the new API, converting a module to the
         new API should happen in a separate commit

The module already consistently uses the new API. I just mistakenly added an import from old one.

**Review by hannes-ucsc — DISMISSED** [review]
One nit: 

```diff

Index: src/azul/terra.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/terra.py b/src/azul/terra.py
--- a/src/azul/terra.py	(revision fa3db5c97b3ca226862c3b844244581b75381d5d)
+++ b/src/azul/terra.py	(date 1727228818605)
@@ -202,7 +202,7 @@
             self.domain.value,
             self.subdomain,
             self.name,
-            self._prefix_str()
+            self._prefix_str
         ])
 
     def qualify_table(self, table_name: str) -> str:
Index: src/azul/indexer/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/indexer/__init__.py b/src/azul/indexer/__init__.py
--- a/src/azul/indexer/__init__.py	(revision fa3db5c97b3ca226862c3b844244581b75381d5d)
+++ b/src/azul/indexer/__init__.py	(date 1727228818618)
@@ -283,6 +283,7 @@
         prefix = Prefix.parse(prefix) if prefix else None
         return rest, prefix
 
+    @property
     def _prefix_str(self) -> str:
         return '' if self.prefix is None else str(self.prefix)
 
@@ -352,7 +353,7 @@
         >>> s == str(SimpleSourceSpec.parse(s))
         True
         """
-        return f'{self.name}:{self._prefix_str()}'
+        return f'{self.name}:{self._prefix_str}'
 
 
 class SourceJSON(TypedDict):
```

The rest looks good. When you merge this please closely monitor the reindex in `dev` and `anvildev`. As a baseline, please perform the equivalent reindex on `develop` prior to merging this. We should wait with promoting it until I can weigh in during stand-up next week, and if we promote it, it should be promoted separately which implies first promoting the parent commit and following that up with another promotion of the merge commit. The first of these two promotions should be ad-hoc (as opposed to scheduled), happen midweek, and wouldn't involve a reindex. The second would happen on a Friday so that the reindex falls on the weekend.

**Review by achave11-ucsc — COMMENTED** [review]
Looks good 🦭✅

🔎👁️ Integration Test CL item is unchecked, maybe it didn't stick?

Feel free to incorporate the fixup commits for the next round.

**Review by achave11-ucsc — APPROVED** [review]
Approved ✅

**Comment by achave11-ucsc** [comment]
Assigned @dsotirho-ucsc for a second approval on this PR, [as prescribed in Slack](https://ucsc-gi.slack.com/archives/C705Y6G9Z/p1727378403924689).

**Review by dsotirho-ucsc — CHANGES_REQUESTED** [review]
Nit:
- `typing.Self` could be applied in other locations inside `/src/azul/indexer/__init__.py` such as `SimpleSourceSpec.parse`, `SourceRef.from_json`, `SourcedBundleFQID.from_json`, and `Bundle.from_json`.

**Comment by nadove-ucsc** [comment]
As discussed at PL, I'm choosing not to incorporate every potential use of `typing.Self` in this PR to avoid modifying a large number of unrelated files.

**Review by dsotirho-ucsc — APPROVED** [review]
Approved.

**Comment by nadove-ucsc** [comment]
### Security design review

- [x] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

**Comment by achave11-ucsc** [comment]
Removing `do not promote` label, since it's no longer appropriate.

**Pull request merged**